components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,377 @@
+            location_sign: None,
+            location: None,
+            need_location: false,
+        };
+        return Ok(Some(builder.into()));
+    }
+
+    // Find the location sign and the location.
+    let (query_without_location, location_sign, location, need_location) =
+        find_location(conn, query_string)?;
+
+    if let (Some(_), false) = (&location, need_location) {
+        // The location sign does not need the specific location, but user is setting something.
+        return Ok(None);
+    }
+
+    if query_without_location.is_empty() {
+        // No remained query.
+        return Ok(None);
+    }
+
+    // Find the modifiers.
+    let (subject_candidate, pre_modifier, post_modifier) =
+        find_modifiers(conn, &query_without_location)?;
+
+    if !is_subject(conn, &subject_candidate)? {
+        return Ok(None);
+    }
+
+    let builder = SuggestionBuilder {
+        query,
+        subject: &subject_candidate,
+        pre_modifier,
+        post_modifier,
+        location_sign,
+        location,
+        need_location,
+    };
+    Ok(Some(builder.into()))
+}
+
+/// Find the location information from the given query string.
+/// The returned result is consited from tuple of below.
+/// (
+///   String: Query string that is removed found location information.
+///   Option<String>: Location sign found in yelp_location_signs table. If not found, returns None.
+///   Option<String>: Specific location name after location sign. If not found, returns None.
+///   bool: Reflects need_location field in the table.
+/// )
+fn find_location(
+    conn: &Connection,
+    query: &str,
+) -> Result<(String, Option<String>, Option<String>, bool)> {
+    let query_with_spaces = format!(" {} ", query);
+    let mut results: Vec<(usize, usize, i8)> = conn.query_rows_and_then(
+        "
+        SELECT
+          INSTR(:query, ' ' || keyword || ' ') AS sign_index,
+          LENGTH(keyword) AS sign_length,
+          need_location
+        FROM yelp_location_signs
+        WHERE
+          sign_index > 0
+        ORDER BY
+          sign_length DESC
+        LIMIT 1
+        ",
+        named_params! {
+            ":query": &query_with_spaces.to_lowercase(),
+        },
+        |row| -> Result<_> {
+            Ok((
+                row.get::<_, usize>("sign_index")?,
+                row.get::<_, usize>("sign_length")?,
+                row.get::<_, i8>("need_location")?,
+            ))
+        },
+    )?;
+
+    let (sign_index, sign_length, need_location) = if let Some(res) = results.pop() {
+        res
+    } else {
+        return Ok((query.trim().to_string(), None, None, false));
+    };
+    let pre_location = query_with_spaces
+        .get(..sign_index)
+        .map(str::trim)
+        .map(str::to_string)
+        .unwrap_or_default();
+    let location_sign = query_with_spaces
+        .get(sign_index..sign_index + sign_length)
+        .map(str::trim)
+        .unwrap_or_default();
```

Comment:
If `.get(sign_index..sign_index + sign_length)` returns `None`, all the subsequent `map` calls will also return `None`, so no need to call `unwrap_or_default()` at the end because we want `location_sign` to be `Option<String>`. Likewise, you don't need
```rs
        if location_sign.is_empty() {
            None
        } else {
            Some(location_sign.to_string())
        }
```
down below if you do
```rs
    let location_sign = query_with_spaces
        .get(sign_index..sign_index + sign_length)
        .map(str::trim)
        .map(str::to_string);
```
As `location_sign` is exactly of type `Option<String>`. `.unwrap_or_default()` is called for `pre_location` because it wants a `String` rather than `Option<String>` :)

Reply:
Thanks, @ncloudioj!
Yeah, I had thought so. But it seems that if the `get()` gets `""`, it returns `Some(""))`.
For example, if run the following code, it shows `Some("")`.

```
let r = "test".get(0..0);
println!("{:?}", r);
```

As I wanted to handle empty string as `None`, checked it explicitly. If you have better idea, could you tell me, please?

Reply:
Ah, I see. You wanted to handle the empty string case. I think you can attach a `filter()` on it.

```
    let location_sign = query_with_spaces
        .get(sign_index..sign_index + sign_length)
        .filter(|s| !s.is_empty())
        .map(str::trim)
        .map(str::to_string);
```

Reply:
Cool! 
I changed the order of the processing like the below to trim first before filtering, it works fine!
Thank you so much!

```
    let location_sign = query_with_spaces
        .get(sign_index..sign_index + sign_length)
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(str::to_string);
```